stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -223,6 +233,133 @@ class cholesky_scalar : public vari {
+        variRefL_[pos] = new vari(L_A.coeffRef(i, j), false);
+        ++pos;
+      }
+    }
+  }
+  /**
+   * Reverse mode differentiation algorithm using a GPU
+   *
+   * Reference:
+   *
+   * Iain Murray: Differentiation of the Cholesky decomposition, 2016.
+   *
+   */
+  virtual void chain() {
+    using Eigen::MatrixXd;
+    MatrixXd Lbar_(M_, M_);
+    MatrixXd L_(M_, M_);
+    Lbar_.setZero();
+    L_.setZero();
+
+    size_t pos = 0;
+    for (size_type j = 0; j < M_; ++j) {
+      for (size_type i = j; i < M_; ++i) {
+        Lbar_.coeffRef(i, j) = variRefL_[pos]->adj_;
+        L_.coeffRef(i, j) = variRefL_[pos]->val_;
+        ++pos;
+      }
+    }
+
+    matrix_gpu L(L_);
+    matrix_gpu Lbar(Lbar_);
+    int M = M_;
+    int block_size_ = std::max((M / 8), 8);
+    block_size_ = std::min(block_size_, 512);
+    // The following is a GPU implementation of
+    // the chain() function from the cholesky_block
+    // vari class implementation
+    for (int k = M; k > 0; k -= block_size_) {
+      int j = std::max(0, k - block_size_);
+      int k_j_ind = k - j;
+      int m_k_ind = M - k;
+
+      matrix_gpu R(k_j_ind, j);
+      matrix_gpu D(k_j_ind, k_j_ind);
+      matrix_gpu B(m_k_ind, j);
+      matrix_gpu C(m_k_ind, k_j_ind);
+
+      matrix_gpu Rbar(k_j_ind, j);
+      matrix_gpu Dbar(k_j_ind, k_j_ind);
+      matrix_gpu Bbar(m_k_ind, j);
+      matrix_gpu Cbar(m_k_ind, k_j_ind);
+
+      R.sub_block(L, j, 0, 0, 0, k_j_ind, j);
+      D.sub_block(L, j, j, 0, 0, k_j_ind, k_j_ind);
+      B.sub_block(L, k, 0, 0, 0, m_k_ind, j);
+      C.sub_block(L, k, j, 0, 0, m_k_ind, k_j_ind);
+
+      Rbar.sub_block(Lbar, j, 0, 0, 0, k_j_ind, j);
+      Dbar.sub_block(Lbar, j, j, 0, 0, k_j_ind, k_j_ind);
+      Bbar.sub_block(Lbar, k, 0, 0, 0, m_k_ind, j);
+      Cbar.sub_block(Lbar, k, j, 0, 0, m_k_ind, k_j_ind);
+
+      Cbar = Cbar * lower_triangular_inverse(D);
+      Bbar = Bbar - Cbar * R;
+      Dbar = Dbar - transpose(Cbar) * C;
+
+      // the implementation of the symbolic_rev inline function
+      // for the GPU
+      Dbar = transpose(D) * Dbar;
+      Dbar.triangular_transpose<TriangularMapGPU::LowerToUpper>();
+      D = transpose(lower_triangular_inverse(D));
+      Dbar = D * transpose(D * Dbar);
+      Dbar.triangular_transpose<TriangularMapGPU::LowerToUpper>();
+      // end of symbolic_rev inline function
+
+      Rbar = Rbar - transpose(Cbar) * B;
+      Rbar = Rbar - Dbar * R;
+      Dbar = diagonal_multiply(Dbar, 0.5);
+      Dbar.zeros<stan::math::TriangularViewGPU::Upper>();
+
+      Lbar.sub_block(Rbar, 0, 0, j, 0, k_j_ind, j);
+      Lbar.sub_block(Dbar, 0, 0, j, j, k_j_ind, k_j_ind);
+      Lbar.sub_block(Bbar, 0, 0, k, 0, m_k_ind, j);
+      Lbar.sub_block(Cbar, 0, 0, k, j, m_k_ind, k_j_ind);
+    }
+    copy(Lbar_, Lbar);
+    pos = 0;
+    for (size_type j = 0; j < M_; ++j)
+      for (size_type i = j; i < M_; ++i)
+        variRefA_[pos++]->adj_ += Lbar_.coeffRef(i, j);
+  }
+};
+#endif
```

Comment:
Sean as I'm looking over the different methods we have here, it looks like they all pretty much do the same constructor but then have different chain methods. Do you think it's worth/good to refactor these in this PR to just have a single class or keep them separate?

Reply:
So my idea to do this simply would be something like

```cpp
class cholesky : public vari {
 public:
  int M_;
  vari** variRefA_;
  vari** variRefL_;
  cholesky(const Eigen::Matrix<var, -1, -1>& A,
                 const Eigen::Matrix<double, -1, -1>& L_A) {
  ...all share the same constructor...
  }
 void chain_scalar() {code..}
 void chain_block() {code..}
 void chain_cl() {code..}
 virtual void chain() {
  if (M_ < small_num) {
    chain_scalar();
  } else if (M_ > big_num) {
    chain_cl();
  } else {
    chain_block();
  }
 }
};
```

If you've tried it before I'll trust your judgement that there's stuff going on here that would make the above more complicated than I think it is